ott/core/initializers.py

```diff
@@ -0,0 +1,279 @@
+      f_potential = f_potential - jnp.mean(f_potential)
+      f_potential = f_potential if lse_mode else ot_problem.scaling_from_potential(
+          f_potential
+      )
+      return f_potential
+
+
+class SortingInit(SinkhornInitializer):
+
+  def __init__(
+      self,
+      vector_min: Optional[bool] = False,
+      tol: Optional[float] = 1e-2,
+      max_iter: Optional[int] = 100,
+      stop_gradient: Optional[bool] = True
+  ) -> None:
+    """Sorting Init class.
+
+    Args:
+        vector_min (Optional[bool], optional): Use vectorized inner loop if true. Defaults to False.
+        tol (Optional[float], optional): DualSort convergence threshold. Defaults to 1e-2.
+        max_iter (Optional[int], optional): Max DualSort steps. Defaults to 100.
+        stop_gradient (Optional[bool], optional): Do not trace gradient through the initializer. Defaults to True.
+    """
+    super().__init__()
+
+    self.tolerance = tol
+    self.stop_gradient = stop_gradient
+    self.max_iter = max_iter
+    self.update_fn = self.vectorized_update if vector_min else self.coordinate_update
+
+  def vectorized_update(
+      self, f: jnp.ndarray, modified_cost: jnp.ndarray
+  ) -> jnp.ndarray:
+    """Inner loop DualSort Update.
+
+    Args:
+        f (jnp.ndarray): potential f, array of size n
+        modified_cost (jnp.ndarray): cost matrix minus diagonal of cost matrix across each column
+    Returns:
+        jnp.ndarray: updated potential vector, f
+    """
+    f = jnp.min(modified_cost + f[None, :], axis=1)
+    return f
+
+  def coordinate_update(
+      self, f: jnp.ndarray, modified_cost: jnp.ndarray
+  ) -> jnp.ndarray:
+    """Coordinate-wise updates within inner loop.
+
+    Args:
+        f (jnp.ndarray): potential f, array of size n
+        modified_cost (jnp.ndarray): cost matrix minus diagonal of cost matrix across each column
+    Returns:
+        jnp.ndarray: updated potential vector, f
+    """
+
+    def body_fn(i, f):
+      new_f = jnp.min(modified_cost[i, :] + f)
+      f = f.at[i].set(new_f)
+      return f
+
+    return jax.lax.fori_loop(0, len(f), body_fn, f)
+
+  def init_sorting_dual(
+      self, modified_cost: jnp.ndarray, f_potential: jnp.ndarray
+  ) -> jnp.ndarray:
+    """Run DualSort algorithm.
+
+    Args:
+        modified_cost (jnp.ndarray): cost matrix minus diagonal of cost matrix across each column
+        f_potential (jnp.ndarray): potential f, array of size n
+    Returns:
+        jnp.ndarray: potential f, array of size n
+    """
+    it = 0
+    diff = self.tolerance + 1.0
+
+    state = (f_potential, diff, it)
+
+    def body_fn(state):
+      prev_f, _, it = state
+      f_potential = self.update_fn(prev_f, modified_cost)
+      diff = jnp.sum((f_potential - prev_f) ** 2)
+      it += 1
+      return f_potential, diff, it
+
+    def cond_fn(state):
+      _, diff, it = state
+      return (diff > self.tolerance) & (it < self.max_iter)
```

Comment:
`&` is a bitwise `and`, use `jnp.logical_and`.

Reply:
done